content/advent-2016/saga-go-dependency-management.md

```diff
@@ -0,0 +1,106 @@
+
+## The road so far
+
+Anyone who takes a stroll down the memory lane of Go package management will find ample clumps of ripped-out hair strewn about. The process has been long and frustrating, and there's far too much to cover in detail here. But the history is important to where we are now, so a high-level timeline is worth it.
+
+`go get` was, and is, the only official tooling for retrieving Go code and placing it on disk. It has served Go users and developers alike since the release of Go 1.0. Hand-in-hand with `go get` has been the general recommendation from the Go Team on how to keep everything working well: "write backwards-compatible code." That's a slightly-rephrased version of the Go 1 guarantee. This pairs well with `go get`, as `go get`'s behavior could only be correct if the following were true:
+
+1. All code is always backwards compatible
+2. There are no fixed points/releases
+
+_(to be clear, the authors of `go get` do not necessarily believe this is good or right - just that when these things aren't true, it's not `go get`'s problem.)_
+
+In an ecosystem structured around `go get`, the most obvious problem was reproducibility: because `go get` fetches the latest code for any packages not already on disk, it's impossible to guarantee that your users  - or your teammates, or build system - will have the same versions of your software's dependencies that you do. Being the most gaping wound, reproducibility was the main problem that tooling has sought to address, especially early on. [godep](https://github.com/tools/godep) - still the most widely used tool today - emerged just thirteen months after the 1.0 release of Go, with a laser focus on reproducible builds.
+
+`go get` had another noteworthy side effect on the Go mindset: most developers don't make releases of their code. `go get` works directly with git, hg, bzr, and svn repositories and deals exclusively with a repository's "default branch." This behavior makes releases irrelevant in the default case. And when tagging and releasing your code makes no difference to the average user, there's not much incentive to do it.
+
+Over time, [many tools have emerged](https://github.com/golang/go/wiki/PackageManagementTools). Some do more, up to replacing the entire build toolchain. Others might proudly describe themselves as little more than `rsync` wrappers. Divisions widened as tools proliferated, but there was at least one generally clear trend: developers wanted to be able to encapsulate dependencies on a project-by-project basis.
+
+The Go team responded to this trend: in Go 1.5, [the `vendor` directory](https://docs.google.com/document/d/1Bz5-UB7g2uPBdOx-rw5t9MxJwkfpx90cqG9AFL0JAYo/edit) was introduced, providing an official, toolchain-supported mechanism by which we could encapsulate a project's dependencies. But it wasn't until February 2016 - when Go 1.6 was released and `vendor` was enabled by default - that the ball really started rolling.
```

Comment:
In general, the piece avoids explanations of mechanics in favor of maintaining a high-level narrative. Mechanics are left to the linked material.